src/libraries/Microsoft.PowerFx.Core/Syntax/UserDefinitions.cs

```diff
@@ -155,5 +166,121 @@ private bool CheckReturnType(IdentToken returnType, List<TexlError> errors)
+        private ParseUserDefinitionResult ProcessPartialAttributes(ParseUserDefinitionResult parsed)
+        {
+            var groupedFormulas = parsed.NamedFormulas.GroupBy(nf => nf.Ident.Name.Value);
+            var errors = parsed.Errors?.ToList() ?? new List<TexlError>();
+            var newFormulas = new List<NamedFormula>();
+
+            foreach (var nameGroup in groupedFormulas)
+            {
+                var name = nameGroup.Key;
+                var firstAttribute = nameGroup.Select(nf => nf.Attribute).FirstOrDefault(att => att != null);
+
+                if (firstAttribute == null || nameGroup.Count() == 1)
+                {
+                    newFormulas.AddRange(nameGroup);
+                    continue;
+                }
+
+                if (firstAttribute.AttributeOperation == PartialAttribute.AttributeOperationKind.Error)
+                {
+                    errors.Add(new TexlError(firstAttribute.AttributeOperationToken, DocumentErrorSeverity.Severe, TexlStrings.ErrUnknownPartialOp));
+                    newFormulas.AddRange(nameGroup);
+                    continue;
+                }
+
+                var updatedGroupFormulas = new List<NamedFormula>();
+                var id = 0;
+                foreach (var formula in nameGroup)
+                {
+                    // This is just for the prototype, since we only have the one kind.
+                    if (formula.Attribute.AttributeName.Name != "Partial")
+                    {
+                        errors.Add(new TexlError(formula.Attribute.AttributeOperationToken, DocumentErrorSeverity.Severe, TexlStrings.ErrOnlyPartialAttribute));
+                        continue;
+                    }
+
+                    if (!firstAttribute.SameAttribute(formula.Attribute))
+                    {
+                        errors.Add(new TexlError(formula.Attribute.AttributeOperationToken, DocumentErrorSeverity.Severe, TexlStrings.ErrOperationDoesntMatch));
+                        continue;
+                    }
+
+                    var newName = new IdentToken(name + _renamedFormulaGuid + id, formula.Ident.Span, isNonSourceIdentToken: true);
+                    id++;
+                    updatedGroupFormulas.Add(new NamedFormula(newName, formula.Formula, formula.StartingIndex, formula.Attribute));
+                }
+
+                if (updatedGroupFormulas.Count != nameGroup.Count())
+                {
+                    // Not all matched, don't use renamed formulas
+                    newFormulas.AddRange(nameGroup);
+                    continue;
+                }
+
+                newFormulas.AddRange(updatedGroupFormulas);
+                newFormulas.Add(
+                    new NamedFormula(
+                        new IdentToken(name, firstAttribute.AttributeName.Span, isNonSourceIdentToken: true),
+                        GetPartialCombinedFormula(name, firstAttribute.AttributeOperation, updatedGroupFormulas),
```

Comment:
>GetPartialCombinedFormula [](http://example.com/codeflow?start=24&length=25)

Does it work with the NF edit optimization that we have added?

Reply:
Mostly, yes, although optimization isn't the focus of this prototype.

Reply:
This is fine for prototype but when we productize it, would this design work well with optimization? As long as you have thought about it and there is a way, it's fine.